src/index.js

```diff
@@ -1,37 +1,43 @@
-function log(prevState, action, nextState) {
-  console.group('%c action', 'color: gray; font-weight: lighter;', action.name);
-  console.log('%c prev state', 'color: #9E9E9E; font-weight: bold;', prevState);
-  console.log('%c data', 'color: #03A9F4; font-weight: bold;', action.data);
-  console.log('%c next state', 'color: #4CAF50; font-weight: bold;', nextState);
-  console.groupEnd();
-}
+import defaultLog from './defaultLog'
 
 export default function(options) {
-  options = options || {};
-  options.log = typeof options.log === 'function' ? options.log : log;
+  options = options || {}
+  options.log = typeof options.log === "function" ? options.log : defaultLog
 
-  return function(emit) {
-    var actionStack = [];
-    return {
-      events: {
-        action: function(state, actions, action) {
-          actionStack.push(action);
-        },
-        resolve: function(state, actions, result) {
-          if (typeof result === 'function') {
-            var action = actionStack.pop();
-            return function(update) {
-              return result(function(result) {
-                actionStack.push(action);
-                return update(result);
-              });
-            };
-          }
-        },
-        update: function(state, actions, nextState) {
-          options.log(state, actionStack.pop(), nextState);
-        }
+  return function(app) {
+    return function(props) {
+      function enhance(actions) {
+        return Object.keys(actions).reduce(function(otherActions, name) {
+          var action = actions[name]
+          otherActions[name] =
+            typeof action === "function"
+              ? function(state, actions, data) {
+                  var result = action(state, actions, data)
+                  if (typeof result === "function") {
+                    return function(update) {
+                      return result(function(withState) {
+                        options.log(
+                          state,
+                          { name: name, data: data },
+                          withState
+                        )
+                        return update(withState)
+                      })
+                    }
+                  } else {
+                    options.log(state, { name: name, data: data }, result)
+                    return result
+                  }
+                }
+              : enhance(action)
+          return otherActions
+        }, {})
       }
-    };
-  };
+
+      props.actions = enhance(props.actions)
+      var appActions = app(props)
```

Comment:
the code is not working with modules

Reply:
@okwolf props.actions doesn't include actions from modules before app(props)

Reply:
@AlexanderChen1989 good catch! I'm new to `modules` and hadn't added a test for them yet

Reply:
I need to remember to include this magic in [`hyperapp-middleware`](https://github.com/okwolf/hyperapp-middleware) also